File: src/data/status.ts
import { execSync } from 'child_process';
import { existsSync, readFileSync } from 'fs';
import { join } from 'path';
import { homedir } from 'os';
import type { ServiceStatus } from '../types.js';

const HEALTH_LOG = join(homedir(), '.openclaw', 'logs', 'config-health.json');

export function loadStatus(): ServiceStatus {
  let active: ServiceStatus['active'] = 'unknown';
  let since = '';
  let socketHealth: ServiceStatus['socketHealth'] = 'unknown';
  let version = '';

  // Active state
  try {
    const out = execSync(
      'systemctl --user is-active openclaw-gateway.service 2>/dev/null',
      { encoding: 'utf-8', timeout: 2000 }
    ).trim();
    active = out === 'active' ? 'running' : out === 'failed' ? 'failed' : 'stopped';
  } catch {
    active = 'stopped';
  }

  // Since when (only if running)
  if (active === 'running') {
    try {
      const out = execSync(
        'systemctl --user show openclaw-gateway.service --property=ActiveEnterTimestamp 2>/dev/null',
        { encoding: 'utf-8', timeout: 2000 }
      ).trim();
      const match = out.match(/ActiveEnterTimestamp=(.+)/);
      if (match?.[1] && match[1] !== 'n/a' && match[1].trim()) {
        const d = new Date(match[1].trim());
        if (!isNaN(d.getTime())) {
          since = d.toLocaleString('en-US', {
            month: 'short', day: 'numeric',
            hour: '2-digit', minute: '2-digit',
          });
        }
      }
    } catch { /* ignore */ }
  }

  // Socket health: check config-health.json first, then journal
  if (existsSync(HEALTH_LOG)) {
    try {
      const health = JSON.parse(readFileSync(HEALTH_LOG, 'utf-8'));
      if (health.socketStatus) {
        socketHealth = health.socketStatus === 'connected' ? 'ok' : 'stale';
      }
    } catch { /* ignore */ }
  }

  if (socketHealth === 'unknown' && active === 'running') {
    try {
      const logs = execSync(
        'journalctl --user -u openclaw-gateway.service -n 20 --no-pager --output=cat 2>/dev/null',
        { encoding: 'utf-8', timeout: 2000 }
      );
      socketHealth = logs.includes('stale-socket') ? 'stale' : 'ok';
    } catch {
      socketHealth = 'ok';
    }
  }

  // Version from npm global package.json (faster than running openclaw --version)
  try {
    const npmRoot = execSync('npm root -g 2>/dev/null', { encoding: 'utf-8', timeout: 2000 }).trim();
    if (npmRoot) {
      const pkgPath = join(npmRoot, 'openclaw', 'package.json');
      if (existsSync(pkgPath)) {
        const pkg = JSON.parse(readFileSync(pkgPath, 'utf-8'));
        version = pkg.version ?? '';
      }
    }
  } catch { /* ignore */ }

  return { active, since, socketHealth, version };
}

File: internal/data/status.go
package data

import (
	"encoding/json"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

func LoadStatus() ServiceStatus {
	var s ServiceStatus
	s.Active = "unknown"
	s.SocketHealth = "unknown"

	// Active state
	out, err := runCmd(2*time.Second, "systemctl", "--user", "is-active", "openclaw-gateway.service")
	if err != nil {
		s.Active = "stopped"
	} else {
		switch strings.TrimSpace(out) {
		case "active":
			s.Active = "running"
		case "failed":
			s.Active = "failed"
		default:
			s.Active = "stopped"
		}
	}

	// Since (only when running)
	if s.Active == "running" {
		out, err := runCmd(2*time.Second, "systemctl", "--user", "show",
			"openclaw-gateway.service", "--property=ActiveEnterTimestamp")
		if err == nil {
			if idx := strings.Index(out, "="); idx >= 0 {
				ts := strings.TrimSpace(out[idx+1:])
				if ts != "" && ts != "n/a" {
					if d, err := time.Parse("Mon 2006-01-02 15:04:05 MST", ts); err == nil {
						s.Since = d.Format("Jan 2, 3:04 PM")
					}
				}
			}
		}
	}

	// Socket health
	home, _ := os.UserHomeDir()
	healthPath := filepath.Join(home, ".openclaw", "logs", "config-health.json")
	if b, err := os.ReadFile(healthPath); err == nil {
		var h struct {
			SocketStatus string `json:"socketStatus"`
		}
		if json.Unmarshal(b, &h) == nil && h.SocketStatus != "" {
			if h.SocketStatus == "connected" {
				s.SocketHealth = "ok"
			} else {
				s.SocketHealth = "stale"
			}
		}
	}
	if s.SocketHealth == "unknown" && s.Active == "running" {
		logs, err := runCmd(2*time.Second, "journalctl", "--user",
			"-u", "openclaw-gateway.service", "-n", "20", "--no-pager", "--output=cat")
		if err == nil {
			if strings.Contains(logs, "stale-socket") {
				s.SocketHealth = "stale"
			} else {
				s.SocketHealth = "ok"
			}
		}
	}

	// Version
	npmOut, err := runCmd(2*time.Second, "npm", "root", "-g")
	if err == nil {
		pkgPath := filepath.Join(strings.TrimSpace(npmOut), "openclaw", "package.json")
		if b, err := os.ReadFile(pkgPath); err == nil {
			var pkg struct {
				Version string `json:"version"`
			}
			if json.Unmarshal(b, &pkg) == nil {
				s.Version = pkg.Version
			}
		}
	}

	return s
}

func runCmd(timeout time.Duration, name string, args ...string) (string, error) {
	cmd := exec.Command(name, args...)
	ch := make(chan struct{ out string; err error }, 1)
	go func() {
		out, err := cmd.Output()
		ch <- struct{ out string; err error }{string(out), err}
	}()
	select {
	case r := <-ch:
		return r.out, r.err
	case <-time.After(timeout):
		cmd.Process.Kill()
		return "", os.ErrDeadlineExceeded
	}
}
